File: modules/admin_database.py
"""
GASMAN – ENTERPRISE DATABASE OPERATIONS MODULE

Features
✔ Monitor gasman + data_logger databases
✔ Table size monitoring
✔ Disk usage monitoring
✔ Old log cleanup
✔ Data logger cleanup
✔ Largest table detection
✔ Table optimization
✔ Production-safe SQL queries
"""

from fastapi import APIRouter, Depends, Body
from config.db_pool import get_gasman_db
from modules.auth_dependency import require_admin
import shutil
import subprocess
from datetime import datetime
from pathlib import Path
from fastapi.responses import FileResponse
import os
SERVICE_PASSWORD = os.getenv("SERVICE_PASSWORD","service_password")

router = APIRouter()
# ----------------------------------------------------------
# BACKUP DIRECTORY (UNIVERSAL PATH)
# ----------------------------------------------------------

BASE_DIR = Path(__file__).resolve().parents[1]
BACKUP_DIR = BASE_DIR / "backups"

BACKUP_DIR.mkdir(parents=True, exist_ok=True)
# ==========================================================
# DATABASE + DISK STATS
# ==========================================================

@router.get("/admin/database-stats")
def database_stats(_: dict = Depends(require_admin)):

    conn = get_gasman_db()
    cur = conn.cursor(dictionary=True)

    databases = ["gasman", "data_logger"]

    database_results = []
    total_size = 0

    try:

        for db in databases:

            # --------------------------------------------------
            # TABLE SIZES
            # --------------------------------------------------

            cur.execute("""
                SELECT
                    table_name,
                    IFNULL(table_rows,0) AS table_rows,
                    ROUND(
                        (IFNULL(data_length,0) + IFNULL(index_length,0))
                        /1024/1024,
                        2
                    ) AS size_mb
                FROM information_schema.tables
                WHERE table_schema = %s
                ORDER BY size_mb DESC
            """, (db,))

            tables = cur.fetchall()

            # --------------------------------------------------
            # DATABASE TOTAL SIZE
            # --------------------------------------------------

            cur.execute("""
                SELECT
                    ROUND(
                        SUM(
                            IFNULL(data_length,0) +
                            IFNULL(index_length,0)
                        ) /1024/1024,
                        2
                    ) AS db_size
                FROM information_schema.tables
                WHERE table_schema = %s
            """, (db,))

            result = cur.fetchone()
            size = result["db_size"] if result["db_size"] else 0

            total_size += size

            database_results.append({
                "name": db,
                "size": size,
                "tables": tables
            })

    finally:
        cur.close()
        conn.close()

    # --------------------------------------------------
    # SERVER DISK USAGE
    # --------------------------------------------------

    total, used, free = shutil.disk_usage("/")

    return {
        "databases": database_results,
        "db_total": round(total_size, 2),
        "disk_total": round(total / 1024 / 1024 / 1024, 2),
        "disk_used": round(used / 1024 / 1024 / 1024, 2),
        "disk_free": round(free / 1024 / 1024 / 1024, 2)
    }


# ==========================================================
# OLD GASMAN LOG COUNT
# ==========================================================

@router.get("/admin/old-log-count")
def old_log_count(_ = Depends(require_admin)):

    conn = get_gasman_db()
    cur = conn.cursor()

    try:

        cur.execute("""
            SELECT COUNT(*)
            FROM gasman_task_activity
            WHERE created_at < NOW() - INTERVAL 90 DAY
        """)

        count = cur.fetchone()[0]

    finally:
        cur.close()
        conn.close()

    return {"old_logs": count}


# ==========================================================
# CLEAN GASMAN LOGS
# ==========================================================

@router.post("/admin/cleanup-gasman")
def cleanup_gasman(_ = Depends(require_admin)):

    conn = get_gasman_db()
    cur = conn.cursor()

    try:

        cur.execute("""
            DELETE FROM gasman_task_activity
            WHERE created_at < NOW() - INTERVAL 90 DAY
        """)

        deleted = cur.rowcount

        conn.commit()

    finally:
        cur.close()
        conn.close()

    return {"deleted": deleted}


# ==========================================================
# CLEAN DATA LOGGER
# ==========================================================

@router.post("/admin/cleanup-datalogger")
def cleanup_datalogger(_ = Depends(require_admin)):

    conn = get_gasman_db()
    cur = conn.cursor()

    raw_deleted = 0
    hist_deleted = 0

    try:

        # --------------------------------------------------
        # RAW TELEMETRY DATA
        # --------------------------------------------------

        cur.execute("""
            DELETE FROM data_logger.raw_table
        """)
        raw_deleted = cur.rowcount

        # --------------------------------------------------
        # HISTORICAL DEVICE LOGS
        # --------------------------------------------------

        cur.execute("""
            DELETE FROM data_logger.device_log_historical
        """)
        hist_deleted = cur.rowcount

        conn.commit()

    except Exception as e:

        conn.rollback()

        return {
            "raw_deleted": 0,
            "historical_deleted": 0,
            "error": str(e)
        }

    finally:
        cur.close()
        conn.close()

    return {
        "raw_deleted": raw_deleted,
        "historical_deleted": hist_deleted
    }


# ==========================================================
# LARGEST TABLE ANALYTICS
# ==========================================================

@router.get("/admin/database-operations")
def database_operations(_ = Depends(require_admin)):

    conn = get_gasman_db()
    cur = conn.cursor(dictionary=True)

    try:

        cur.execute("""
            SELECT
                table_schema,
                table_name,
                IFNULL(table_rows,0) AS table_rows,
                ROUND(
                    (IFNULL(data_length,0) +
                     IFNULL(index_length,0))
                    /1024/1024,
                    2
                ) AS size_mb
            FROM information_schema.tables
            WHERE table_schema IN ('gasman','data_logger')
            ORDER BY size_mb DESC
            LIMIT 50
        """)

        tables = cur.fetchall()

    finally:
        cur.close()
        conn.close()

    return {"largest_tables": tables}


# ==========================================================
# OPTIMIZE TABLES
# ==========================================================

@router.post("/admin/optimize-tables")
def optimize_tables(_ = Depends(require_admin)):

    conn = get_gasman_db()
    cur = conn.cursor()

    optimized = 0

    try:

        cur.execute("""
            SELECT table_schema, table_name
            FROM information_schema.tables
            WHERE table_schema IN ('gasman','data_logger')
        """)

        tables = cur.fetchall()

        for schema, table in tables:

            try:
                cur.execute(f"OPTIMIZE TABLE {schema}.{table}")
                optimized += 1
            except:
                # skip tables that fail
                pass

        conn.commit()

    finally:
        cur.close()
        conn.close()

    return {
        "status": "optimized",
        "tables_optimized": optimized
    }
# ==========================================================
# CREATE BACKUP
# ==========================================================

@router.post("/admin/create-backup")
def create_backup(_ = Depends(require_admin)):

    timestamp = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")

    filename = f"gasman_backup_{timestamp}.sql"

    filepath = BACKUP_DIR / filename

    cmd = [
        "mysqldump",
        "-u", "root",
        "-proot",
        "gasman",
        "data_logger"
    ]

    with open(filepath, "w") as f:
        subprocess.run(cmd, stdout=f, check=True)

    return {
        "status": "success",
        "file": filename
    }


# ==========================================================
# LIST BACKUPS
# ==========================================================

@router.get("/admin/list-backups")
def list_backups(_ = Depends(require_admin)):

    files = []

    for path in sorted(BACKUP_DIR.glob("*.sql"), reverse=True):

        files.append({
            "name": path.name,
            "size": round(path.stat().st_size / 1024 / 1024, 2),
            "created": datetime.fromtimestamp(path.stat().st_ctime)
        })

    return {"backups": files}
# ==========================================================
# DOWNLOAD BACKUP
# ==========================================================

@router.get("/admin/download-backup/{file}")
def download_backup(file: str, _ = Depends(require_admin)):

    path = BACKUP_DIR / file

    return FileResponse(
        path,
        filename=file,
        media_type="application/sql"
    )


# ==========================================================
# RESTORE BACKUP
# ==========================================================

@router.post("/admin/restore-backup/{file}")
def restore_backup(file: str, _ = Depends(require_admin)):

    path = BACKUP_DIR / file

    cmd = [
        "mysql",
        "-u", "root",
        "-proot"
    ]

    with open(path, "r") as f:
        subprocess.run(cmd, stdin=f, check=True)

    return {"status": "restored"}
# ==========================================================
# DELETE BACKUP
# ==========================================================

@router.delete("/admin/delete-backup/{file}")
def delete_backup(file: str, _ = Depends(require_admin)):

    try:

        path = BACKUP_DIR / file

        if not path.exists():
            return {"status": "error", "message": "Backup not found"}

        path.unlink()

        return {"status": "deleted", "file": file}

    except Exception as e:

        return {"status": "error", "message": str(e)}
# ==========================================================
# VERIFY SERVICE PASSWORD
# ==========================================================

from fastapi import Body

@router.post("/admin/verify-service-password")
def verify_service_password(data: dict = Body(...), _ = Depends(require_admin)):

    password = data.get("password")

    if password == SERVICE_PASSWORD:
        return {"status": "ok"}

    return {"status": "invalid"}

File: static/admin/admin_database.js
/* =========================================================
   GASMAN ENTERPRISE DATABASE OPERATIONS PANEL
   Frontend Controller

   Features
   ✔ Database KPI monitoring
   ✔ Largest tables analytics
   ✔ GASMAN log cleanup
   ✔ DATA LOGGER cleanup
   ✔ Table optimization
   ✔ Safe API handling
   ✔ Auto refresh
========================================================= */


/* =========================================================
   UTILITY SAFE ELEMENT
========================================================= */

function el(id){
return document.getElementById(id);
}


/* =========================================================
   LOAD DATABASE STATS
========================================================= */

async function loadDatabaseStats(){

try{

const r = await fetch("/admin/database-stats");

if(!r.ok) throw new Error("API error");

const data = await r.json();

/* KPI VALUES */

if(el("dbSize"))
el("dbSize").innerText =
(Number(data.db_total)||0).toFixed(2) + " MB";

if(el("diskTotal"))
el("diskTotal").innerText =
(Number(data.disk_total)||0).toFixed(2) + " GB";

if(el("diskUsed"))
el("diskUsed").innerText =
(Number(data.disk_used)||0).toFixed(2) + " GB";

if(el("diskFree"))
el("diskFree").innerText =
(Number(data.disk_free)||0).toFixed(2) + " GB";

}catch(e){

console.error("Database stats failed:",e);

}

}


/* =========================================================
   LOAD OLD GASMAN LOG COUNT
========================================================= */

async function loadOldLogs(){

try{

const r = await fetch("/admin/old-log-count");

if(!r.ok) throw new Error("API error");

const data = await r.json();

if(el("kpiOldLogs"))
el("kpiOldLogs").innerText =
data.old_logs || 0;

}catch(e){

console.error("Old log count failed:",e);

}

}


/* =========================================================
   CLEAN GASMAN LOGS
========================================================= */

async function cleanupGasman(){
const ok = await verifyServicePassword();
if(!ok) return;
if(!confirm("Delete GASMAN logs older than 90 days?")) return;

try{

const r = await fetch("/admin/cleanup-gasman",{method:"POST"});

if(!r.ok) throw new Error("API error");

const data = await r.json();

alert("Deleted " + (data.deleted || 0) + " GASMAN logs");

loadDatabaseStats();
loadOldLogs();
loadLargestTables();

}catch(e){

console.error(e);
alert("Cleanup failed");

}

}


/* =========================================================
   CLEAN DATA LOGGER TABLES
========================================================= */

async function cleanupLogger(){
const ok = await verifyServicePassword();
if(!ok) return;
if(!confirm("Clean DATA LOGGER tables? This will remove telemetry history.")) return;

try{

const r = await fetch("/admin/cleanup-datalogger",{method:"POST"});

if(!r.ok) throw new Error("API error");

const data = await r.json();

if(data.error){

alert("Cleanup error: " + data.error);
return;

}

alert(
"Cleanup complete\n\n" +
"Raw rows deleted: " + (data.raw_deleted || 0) +
"\nHistorical rows deleted: " + (data.historical_deleted || 0)
);

loadDatabaseStats();
loadLargestTables();

}catch(e){

console.error(e);
alert("Cleanup failed");

}

}


/* =========================================================
   LOAD LARGEST TABLE ANALYTICS
========================================================= */

async function loadLargestTables(){

try{

const r = await fetch("/admin/database-operations");

if(!r.ok) throw new Error("API error");

const data = await r.json();

const body = el("largestTables");

if(!body) return;

if(!data.largest_tables || !data.largest_tables.length){

body.innerHTML = `
<tr>
<td colspan="4" class="text-center text-muted">
No tables found
</td>
</tr>
`;

return;

}

body.innerHTML = data.largest_tables.map(t=>{

const db =
t.table_schema ??
t.TABLE_SCHEMA ??
"unknown";

const table =
t.table_name ??
t.TABLE_NAME ??
"unknown";

const rows =
Number(
t.table_rows ??
t.TABLE_ROWS ??
0
);

const size =
Number(
t.size_mb ??
0
);

return `
<tr>
<td>${db}</td>
<td>${table}</td>
<td>${rows}</td>
<td>${size.toFixed(2)}</td>
</tr>
`;

}).join("");

}catch(e){

console.error("Largest tables load failed:",e);

}

}


/* =========================================================
   OPTIMIZE ALL TABLES
========================================================= */

async function optimizeTables(){
const ok = await verifyServicePassword();
if(!ok) return;
if(!confirm("Optimize all database tables?\n\nThis may take some time.")) return;

try{

const r = await fetch("/admin/optimize-tables",{method:"POST"});

if(!r.ok) throw new Error("API error");

const data = await r.json();

alert(
"Optimization complete\n\nTables optimized: " +
(data.tables_optimized || 0)
);

loadDatabaseStats();
loadLargestTables();

}catch(e){

console.error(e);
alert("Optimization failed");

}

}
/* =========================================================
   CREATE BACKUP
========================================================= */

async function createBackup(){
const ok = await verifyServicePassword();
if(!ok) return;
if(!confirm("Create database backup?")) return;

const r = await fetch("/admin/create-backup",{method:"POST"});

const data = await r.json();

alert("Backup created: " + data.file);

loadBackups();

}


/* =========================================================
   LOAD BACKUPS
========================================================= */

async function loadBackups(){

const r = await fetch("/admin/list-backups");

const data = await r.json();

const table = document.getElementById("backupTable");

table.innerHTML = data.backups.map(b=>`

<tr>
<td>${b.name}</td>
<td>${b.size} MB</td>
<td>${new Date(b.created).toLocaleString()}</td>

<td>

<button class="btn btn-sm btn-success"
onclick="downloadBackup('${b.name}')">

Download

</button>

<button class="btn btn-sm btn-danger"
onclick="restoreBackup('${b.name}')">

Restore

</button>

<button class="btn btn-sm btn-danger"
onclick="deleteBackup('${b.name}')">

Delete

</button>
</td>

</tr>

`).join("");

}


/* =========================================================
   DOWNLOAD BACKUP
========================================================= */

function downloadBackup(file){

window.location =
"/admin/download-backup/" + file;

}


/* =========================================================
   RESTORE BACKUP
========================================================= */

async function restoreBackup(file){
const ok = await verifyServicePassword();
if(!ok) return;
if(!confirm("Restore backup?\nThis will overwrite current database.")) return;

await fetch("/admin/restore-backup/"+file,{
method:"POST"
});

alert("Database restored");

}

/* =========================================================
   DELETE BACKUP
========================================================= */

async function deleteBackup(file){
const ok = await verifyServicePassword();
if(!ok) return;

if(!confirm("Delete backup?\n\n" + file)) return;

try{

const r = await fetch("/admin/delete-backup/" + file,{
method:"DELETE"
});

const data = await r.json();

if(data.status === "deleted"){

alert("Backup deleted");

loadBackups();

}else{

alert("Delete failed: " + data.message);

}

}catch(e){

alert("Delete failed");

}

}
/* =========================================================
   INITIAL DASHBOARD LOAD
========================================================= */

function initDatabaseDashboard(){

loadDatabaseStats();
loadOldLogs();
loadLargestTables();

}

/* =========================================================
   SERVICE PASSWORD MODAL CONTROLLER
========================================================= */

let serviceResolver = null;
let serviceModal = null;

function verifyServicePassword(){

serviceModal = new bootstrap.Modal(
document.getElementById("servicePasswordModal")
);

document.getElementById("servicePasswordInput").value = "";
document.getElementById("servicePasswordError").style.display = "none";

serviceModal.show();

return new Promise(resolve=>{
serviceResolver = resolve;
});

}


async function submitServicePassword(){

const password =
document.getElementById("servicePasswordInput").value;

const r = await fetch("/admin/verify-service-password",{
method:"POST",
headers:{
"Content-Type":"application/json"
},
body:JSON.stringify({password})
});

const data = await r.json();

if(data.status !== "ok"){

document.getElementById("servicePasswordError").style.display = "block";
return;

}

serviceModal.hide();

if(serviceResolver){
serviceResolver(true);
}

}
/* =========================================================
   AUTO REFRESH
========================================================= */

setInterval(()=>{

loadDatabaseStats();
loadOldLogs();
loadLargestTables();

},60000); // refresh every 60 seconds


/* =========================================================
   START DASHBOARD
========================================================= */

initDatabaseDashboard();
